### PR TITLE
Fix playback when skipping an in progress recording to real-time

### PR DESCRIPTION
**Pull request by jahutchi**
## Description

This PR resolves issue #379: HTSPVFS and in progress recordings

## Motivation and Context

Currently, if you skip right to the end of an in progress recording, then the playback completely chokes; i.e. both the audio and video break up badly, the video freezes, and the audio makes popping noises.

On debugging, I found that the main issue was that Tvheadend can briefly return 0 bytes when playing an in-progress recording towards EOF; the demuxer expects at least 1 bytes to be returned.

My PR resolves this issue by adjusting HTSPVFS::Read() to retry when 0 bytes are returned for an in-progress recording.

After fixing this problem, the other issue I found was that Kodi would buffer for around 10 seconds when skipping fwd to real-time, before eventually resuming normal playback. After some debugging and research I landed on the IsRealTimeStream API call. This PR returns true for IsRealTimeStream when playback is within the last 10 seconds of an in-progress recording.

## How Has This Been Tested?

My testing was over a Gigabit Ethernet using both HD and SD channels.

Backend: TVheadend (v4.2.8) running on a 10yr old HTPC (Acer Revo 3700) running Debian Stretch
Kodi: Running on a Vero4K+ using OSMC Leia nightlies with pvr.hts v4.4.16 (+ this patch).

In this setup, I have conducted several tests as follows:

* Watched a normal, completed recording from start to finish to ensure no regressions were introduced for normal recordings. Also added extra debug logging to show the value of IsRealTimeStream, and could see it was always set to false as it always has been.
* Played an in-progress recording and skipped way beyond real-time playback. This caused the seek to 'snap back' to the end of file i.e. "Real-time" playback.  Playback started almost immediately (buffering circle appeared for only a split second), and playback was fine thereafter.
* Played an in-progress recording (1hr long) and skipped to real-time playback. Then watched the recording right to the end. No stutters or playback issues were observed, and when the show finished the playback ended immediately.
* With extra debugging in place, I monitored the value of IsRealTimeStream on an in-progress recording. I performed several small skip operations, and IsRealTimeStream was only set to true when nearing the end of the file (within last 10 seconds).
* Played LiveTV and also tried skipping backwards and forwards in the timeshift buffer to ensure no regressions were introduced in this area.
* Repeated the tests of skipping an in-progress recording to real-time playback using different chunk sizes. I have tried with both the minimum (4KB) and maximum (512KB) and playback was fine in all cases.
* Tried starting a recording and playing it almost immediately after. Playback was fine and I skipped forward (+2-3 seconds) to real-time playback without any problems / buffering /stuttering.

New tests added (16/04):
* Skip forward to real-time playback, then hit pause for 10 seconds, hit play and watch the logfile to ensure that IsRealTimeStream is set to false
* Skip forward to real-time playback, hit pause, then hit skip back (10 seconds) followed by play. Watch the logfile to ensure that IsRealTimeStream is set back to false.
* Pause an in progress recording (non real-time); whilst paused, skip forward to real-time then hit play. Ensure that IsRealTimeStream is set to true.
* Pause an in progress recording (non real-time); whilst paused, skip forward to real-time, then wait >10 seconds, then hit play. Ensure that IsRealTimeStream is set to false.

### CODE REVIEW

**Comment by ksooo** [comment]
@FernetMenta what is your opinion on this?

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: Wouldn't it be better to calculate with a percentage here instead of a fixed (magic) byte value?

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: @ksooo

Sounds good. To be honest I was struggling to decide on what condition to trigger IsRealTimeStream.

So....what %played of an in-progress recording would you recommend for toggling IsRealTimeStream?

Comment (michaelarnauts) [reply src/tvheadend/HTSPVFS.cpp]: I wouldn't use a percentage. This will make the real-time window much shorter on a new recording and really long on a large recording.

I think fixed is the best, unless we can check the remaining seconds or something.

Comment (ksooo) [reply src/tvheadend/HTSPVFS.cpp]: Right, percentage is not a better approach; best would be 10 seconds till EOF. No idea how to calculate that, though.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: I fully agree that time-based would be best; I originally tried to find ways to calculate this...However it seems in the context of pvr plugin there is no concept of time when dealing with recordings; just filesizes and offsets.

I also wondered whether the tipping point should be a factor of readchunksize; but dont know the VideoPlayer side of things well enough.

Guess the question is...when playing in non-real time mode then how far ahead does it need to be able to read, without delay, in order to avoid buffering. A size, number of seconds, number of readchunksizes?

If it is 10 seconds i think best we can do is agreed on a max filesize which can represent 10s.

I experimented with different sizes and the point it seemed to stop buffering after skipping beyond eof was somewhere between 3-4mb on a hd recording, and around 8mb on uhd (astra demo uhd channel). This is where the 10mb came from. In my tests I also found that simply setting IsRealTimeStream permanently for all recordings works just fine too. Not suggesting we do that but just trying to get things in context.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Looking further at this we can find epoch of when recording started using m_playingRecording->GetFilesStart().

Based on this, current epoch, and current file size, we can work out based on avg bitrate, how many bytes represents 10secs.

Will update PR when done.....

Comment (ksooo) [reply src/tvheadend/HTSPVFS.cpp]: 👍

Comment (FernetMenta) [reply src/tvheadend/HTSPVFS.cpp]: I would use different values for entering realtime and leaving realtime. You don't want the flag to toggle when playing pointer is on the boundary.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: I've adjusted the PR to toggle isRealTimeStream based on whether playback is within 10 seconds (rather than 10mb) of EOF.

It only calculates this as part of Seek operations and stores the bool, so if playing pointer happens to be on the boundary then there's no danger of constantly toggling the flag.

Comment (FernetMenta) [reply src/tvheadend/HTSPVFS.cpp]: > It only calculates this as part of Seek operations

imo this is incomplete. you can pause a recording that is close to live and after a while it should become non realtime

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: @FernetMenta Good spot... clearly will need to re-work this.
I had tried to avoid these calculations after every read when playing an in-progress recording, but maybe that's unavoidable.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Just working on an update to the PR to recalculate m_isRealTimeStream:

1) Following a seek
2) Following each read, but only when m_isRealTimeStream is currently TRUE

If you're in non-realtime mode, then I don't see any way to move into realtime (<10s of EOF), other than a seek (forward) operation. But if you're currently in real-time, then there is obviously the possibility of pause/play to go back to non-realtime (>10s of EOF) which I'd overlooked.

Also, this is the only way I can see us avoiding the danger of constantly toggling the flag when on the boundary during playback.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: I've adjusted the PR as above

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Having tested my previous change I spotted a floor in the logic of triggering the re-calculation on each Read().

During playback kodi (obviously) reads ahead of the play position (updating m_offset as it goes).

Therefore, doing the same calculations (based on m_offset) after each Read() we are actually calculating how far the end of the read buffer (rather than our playback position) is from the end of the file.

The calculation (based on m_offset) at the end of Seek() is valid because we haven't yet done the first Read() following the Seek; thus m_offset correctly reflects the current playback position.

To correctly handle Pause/Play I've therefore made some changes based on the PauseStream() API call; storing the time that the stream was paused (m_timePaused), and later using this when playback resumes to add a number of seconds to the stored offset (m_eofOffsetSecs); which was calculated during previous Seek().

If a Seek() is done whilst paused then the same calculations are performed, and m_timePaused is additionally updated to the current time. Thus when playback resumes then we can update m_eofOffsetSecs to the correct value.

I've added a few extra test cases above to cover "Pause->Play" and "Pause->Seek->Play".

I plan to run through all my tests (above) again later this week, but would value feedback in the meantime.

**Comment by jahutchi** [comment]
I've completed my testing using the tests outlined above, and all tests above passed successfully - now handles pauses correctly WRT setting IsRealTimeStream :-)

File: src/Tvheadend.cpp
Comment (ksooo) [inline src/Tvheadend.cpp]: As you are using `m_playingRecording->GetState() == PVR_TIMER_STATE_RECORDING` a lot, you probably should introduce a private inline member function`CTvheadend::VfsIsActiveRecording() const`.

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: system headers **after** module header, please.

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: line break after m_offset init, please

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp:128]: Maybe you should log a warning when leaving after 50 unsuccesful retries?

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: `time` call is rather expensive. So, just call it once and reuse the value. This is even more accurate.

To put `std::time(nullptr) - m_fileStart` in a variable seems best to me.

And pleaes, use `nullptr` for new code (even if we still use `NULL ` at other places).

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: c++ style cast, please.

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: nullptr

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: nullptr

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: nullptr :-)

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: c++ style cast :-)

**Review by ksooo — CHANGES_REQUESTED** [review]
minors only.

File: src/Tvheadend.cpp
Comment (jahutchi) [reply src/Tvheadend.cpp]: Done - New inline VfsIsActiveRecording() now used in each of the 4 places this check was used.

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Apologies.... now moved

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Done

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp:128]: Done... adjusted break to return, and added a log message after the loop

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Thanks for letting me know... I'm new to contributing to this and the kodi project, so picking up these things as I learn the ropes. Had only gone for time(NULL) as I could see that was how it was done elsewhere in this project.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Done

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Done

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Done

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Done

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Done

**Comment by jahutchi** [comment]
@ksooo care to take another look?

I don't expect these changes to have caused any regressions, but please don't merge until I've retested (tonight) just to be on the safe side.

File: src/Tvheadend.cpp
Comment (ksooo) [reply src/Tvheadend.cpp]: Nope, cannot see this change in the latest commit.

**Comment by ksooo** [comment]
You have not (force-)pushed the changes so far. Cannot re-review.

**Comment by jahutchi** [comment]
> 
> 
> You have not (force-)pushed the changes so far. Cannot re-review.

My apologies.... trying again

File: src/Tvheadend.h
Comment (ksooo) [inline src/Tvheadend.h]: I would add a null check for `m_playingRecording`before accessing it, like `return m_playingRecording && m_playingRecording->GetState() == PVR_TIMER_STATE_RECORDING;`

I know it is not necessary atm, but just to be on the safe side.

File: src/Tvheadend.cpp
Comment (ksooo) [inline src/Tvheadend.cpp]: I would remove this comment, as the following code line should be pretty self-explaining.

**Comment by jahutchi** [comment]
Now ready for re-review  ;-)

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: Well, I meant to put it after `#include "libhts/htsmsg_binary.h"` (but outside the extern C block), please.

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: Is this an error or a warning? I think it is more a warning.

Comment (ksooo) [reply src/tvheadend/HTSPVFS.cpp]: Could't this method be simplified to:

```c++
{
  /* Not opened */
  if (!m_fileId)
    return -1;

  /* Tvheadend may briefly return 0 bytes when playing an in-progress recording at end-of-file
     we'll retry 50 times with 10ms pauses (~500ms) before giving up */
  int tries = inprogress ? 1 : 50;

  for (int i = 1; i <= tries; i++)
  {
    ssize_t read = SendFileRead(buf, len);
    if (read > 0)
    {
      m_offset += read;
      return read;
    }
    else if (i < tries)
      usleep(10000);
  }
  Logger::Log(LogLevel::LEVEL_WARNING, "vfs inprogress recording read failed after %d attempts", tries);
  return -1;
}
```

(not tested!)

**Review by ksooo — CHANGES_REQUESTED** [review]
almost done, I'd say.

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: That's certainly cleaner code... will adjust the PR

Had thought about doing it that way; think I was trying to make the intention of the loop clear. That said, I later added the comment which explains things.

Will adjust the log message too as this makes it (potentially) apply to normal recordings too.

Not too sure about returning -1 when (final) read could be 0 - this is a change in current behaviour.

Shouldn't we return whatever was returned by the previous SendFileRead like it does currently?

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Done

File: src/Tvheadend.h
Comment (jahutchi) [reply src/Tvheadend.h]: Done

File: src/Tvheadend.cpp
Comment (jahutchi) [reply src/Tvheadend.cpp]: Done

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Hows this look

```cpp
{
  /* Not opened */
  if (!m_fileId)
    return -1;

  /* Tvheadend may briefly return 0 bytes when playing an in-progress recording at end-of-file
     we'll retry 50 times with 10ms pauses (~500ms) before giving up */
  int tries = inprogress ? 1 : 50;
  ssize_t read;

  for (int i = 1; i <= tries; i++)
  {
    read = SendFileRead(buf, len);
    if (read > 0)
    {
      m_offset += read;
      return read;
    }
    else if (i < tries)
      usleep(10000);
  }
  Logger::Log(LogLevel::LEVEL_WARNING, "vfs read failed after %d attempts", tries);
  return read;
}
```

Comment (ksooo) [reply src/tvheadend/HTSPVFS.cpp]: Looking good, except the not initialized 'read' variable. If somebody changes the code later so that the loop might be not executed at all we will end with a random return value.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Agreed will start it out at zero.

However, not currently able to compile as LEVEL_WARNING is not declared....

```
pvr.hts/src/tvheadend/HTSPVFS.cpp: In member function 'ssize_t tvheadend::HTSPVFS::Read(unsigned char*, unsigned int, bool)':
pvr.hts/src/tvheadend/HTSPVFS.cpp:127:15: error: 'LEVEL_WARNING' is not a member of 'tvheadend::utilities::LogLevel'
```

Looks like we only have these declared in Logger.h:
      LEVEL_ERROR,
      LEVEL_INFO,
      LEVEL_DEBUG,
      LEVEL_TRACE

Shall I go for LEVEL_DEBUG / INFO or extend "enum LogLevel" in src/tvheadend/utilities/Logger.h to include LEVEL_WARNING?

Comment (ksooo) [reply src/tvheadend/HTSPVFS.cpp]: I suggest to go far DEBUG.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Done

**Comment by jahutchi** [comment]
> 
> 
> almost done, I'd say.

Latest changes pushed... still need to test

**Review by ksooo — APPROVED** [review]
Nice. Tell me, when you're done with your tests.

File: src/tvheadend/HTSPVFS.cpp
Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Just found this obvious mistake so will correct this:
`int tries = inprogress ? 1 : 50;`

Should be:
`int tries = inprogress ? 50 : 1;`

Will update the PR shortly.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Now corrected

File: src/tvheadend/HTSPVFS.cpp
Comment (ksooo) [inline src/tvheadend/HTSPVFS.cpp]: Is usleep available on Windows? If not, we need a portable solution before we can merge this.

Comment (ksooo) [reply src/tvheadend/HTSPVFS.cpp]: We could use https://en.cppreference.com/w/cpp/thread/sleep_for

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: I was just looking at the same page.

This seems to compile
```cpp
#include <chrono>
#include <thread>
...
else if (i < tries)
 std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment (ksooo) [reply src/tvheadend/HTSPVFS.cpp]: yeah, I guess this will do the trick.

Comment (jahutchi) [reply src/tvheadend/HTSPVFS.cpp]: Change pushed....still need to run through my tests...

Comment (ksooo) [reply src/tvheadend/HTSPVFS.cpp]: sure, ping me if you're done

**Comment by jahutchi** [comment]
@ksooo I'm done with my testing, all working as expected.

**Comment by ksooo** [comment]
Here we go... thanks for your contribution.

File: src/client.cpp
Comment (FernetMenta) [inline src/client.cpp:315]: using this function is wrong. you have to recalculate the conditions in read because player can run at lower speed, i.e. 0.8.

**Pull request merged**
